Fee Estimation: Ignore all transactions that are CPFP'd

ismaelsadeeq opened the pull request:
Another attempt at #25380 with an alternate approach

This PR updates `CBlockPolicyEstimator` to ignore all transactions that are CPFP'd by some child when a new block is received.
This fixes the assumption that all transactions are confirmed solely because of their fee rate. 
As some transactions are confirmed due to a CPFP by some child.

This approach linearize all transactions removed from the mempool because of the new  block, and only ignore transactions whose mining score is not the same with their the fee rate.

All transaction with in-mempool parent are already not tracked for fee estimation, so the child that CPFP'd the parent is also ignored.

Upon implementing the cluster mempool, we will just track chunks and make the fee estimator package aware.


This PR depends on #33216, which enables the mini-miner to return the package fee rate of linearized transactions.
That’s why this PR is currently in draft.

darosior committed:
test: keep the list of confirmed utxos up to date in RBF test

ismaelsadeeq committed:
fees: add function that computes ancestors and descendants of txs

- GetTxAncestorsAndDescendants takes the vector of transactions removed from the mempool
  after a block is connected. The function assumed that the order the transactions were
  included in the block was maintained; that is, all transaction parents was be added
  into the vector first before descendants.

- GetTxAncestorsAndDescendants computes all the ancestors and descendants of the transactions.
  The transaction is also included as a descendant and ancestor of itself.

Co-authored-by: willcl-ark <will@256k1.dev>

ismaelsadeeq committed:
fees: ignore all transactions that are CPFP'd

ismaelsadeeq committed:
test: test we ignore parent transactions in a CPFP

Co-authored-by: Antoine Poinsot <darosior@protonmail.com>

glozow commented:
cc @josibake @instagibbs @naumenkogs @darosior @murchandamus since you reviewed the last time

instagibbs commented:
concept ACK

strikes me as the right idea, keep it simple for now, focusing on correctness. Do we have charts anywhere tracking % of transactions that are in a cluster size of 1?

ismaelsadeeq commented:
> Do we have charts anywhere tracking % of transactions that are in a cluster size of 1?

I will analyze the percentage of cluster size 1 transactions mined in previous blocks.

josibake commented:
Concept ACK

rkrux commented on test/functional/feature_fee_estimation.py:
```diff
@@ -383,6 +401,85 @@ def test_acceptstalefeeestimates_option(self):
+        high_feerate = Decimal(20)
+
+        self.log.info("Test that fee estimator will ignore all transaction with in block child")
+        # If a transaction got mined and has a child in the same block it was mined
+        # it does not get accounted in the fee estimator.
+        tx = send_tx(wallet=self.wallet, node=broadcaster, utxo=None, feerate=low_feerate)
+        u = {"txid": tx["txid"], "vout": 0, "value": Decimal(tx["tx"].vout[0].nValue) / COIN}
+        send_tx(wallet=self.wallet, node=broadcaster, utxo=u, feerate=high_feerate)
+        self.sync_mempools(wait=0.1, nodes=[broadcaster, miner])
+        self.generate(miner, 1)
+        assert_equal(broadcaster.estimaterawfee(1)["short"]["fail"]["totalconfirmed"], 0)
+
+        # If it has descendants which have a lower ancestor score, it also does not.
+        tx = send_tx(wallet=self.wallet, node=broadcaster, utxo=None, feerate=high_feerate)
+        u = {"txid": tx["txid"], "vout": 0, "value": Decimal(tx["tx"].vout[0].nValue) / COIN}
+        send_tx(wallet=self.wallet, node=broadcaster, utxo=u, feerate=low_feerate)
+        self.sync_mempools(wait=0.1, nodes=[broadcaster, miner])
+        self.generate(miner, 1)
+        assert_equal(broadcaster.estimaterawfee(1)["short"]["fail"]["totalconfirmed"], 0)
+
+        # Even if it's equal fee rate.
+        tx = send_tx(wallet=self.wallet, node=broadcaster, utxo=None, feerate=high_feerate)
+        u = {"txid": tx["txid"], "vout": 0, "value": Decimal(tx["tx"].vout[0].nValue) / COIN}
+        send_tx(wallet=self.wallet, node=broadcaster, utxo=u, feerate=high_feerate)
+        self.sync_mempools(wait=0.1, nodes=[broadcaster, miner])
+        self.generate(miner, 1)
+        assert_equal(broadcaster.estimaterawfee(1)["short"]["fail"]["totalconfirmed"], 0)
```
This is an easy candidate to remove code duplication by extracting this out in an internal function, and by accepting `tx` and `feeRate` as function args.

ismaelsadeeq replied:
Yes, this is fixed

rkrux commented on test/functional/feature_fee_estimation.py:
```diff
@@ -383,6 +401,85 @@ def test_acceptstalefeeestimates_option(self):
+                txs.append(tx)
+            batch_send_tx = (broadcaster.sendrawtransaction.get_request(tx["hex"]) for tx in txs)
+            broadcaster.batch(batch_send_tx)
+            self.sync_mempools(wait=0.1, nodes=[broadcaster, miner])
```
I've seen this function call in other functional tests as well, can you share why this call is required?

rkrux replied:
Oh nevermind, reviewing this PR gave me the answer: https://github.com/bitcoin/bitcoin/pull/29982

rkrux commented on test/functional/feature_fee_estimation.py:
```diff
@@ -129,6 +129,15 @@ def make_tx(wallet, utxo, feerate):
     )
 
 
+def send_tx(wallet, node, utxo, feerate):
```
Nit: The meaning would be more explicit if we called the last argument as `feerate_satvb`.

ismaelsadeeq replied:
Their is a description below that states that the feerate is in sat/vb so I will leave this as is.

rkrux commented on test/functional/feature_fee_estimation.py:
```diff
@@ -129,6 +129,15 @@ def make_tx(wallet, utxo, feerate):
+    return wallet.send_self_transfer(
+        from_node=node,
+        utxo_to_spend=utxo,
+        fee_rate=Decimal(feerate * 1000) / COIN,
```
I dont fully understand this conversion, is the intention to convert it to `feerate_satkvb`?

rkrux commented on src/policy/fees.cpp:
```diff
@@ -690,6 +690,20 @@ void CBlockPolicyEstimator::processBlock(const std::vector<RemovedMempoolTransac
     shortStats->UpdateMovingAverages();
     longStats->UpdateMovingAverages();
 
+    std::set<Txid> seen_transactions;
+    for (const auto& tx : txs_removed_for_block) {
+        seen_transactions.insert(tx.info.m_tx->GetHash());
+        for (const auto& input : tx.info.m_tx->vin) {
+            const Txid& parentId = input.prevout.hash;
+            if (seen_transactions.count(parentId)) {
+                // Ignore any transaction with a child in the received block
+                // It may be fee bumped by the child.
+                _removeTx(parentId, /*inblock=*/true);
+                seen_transactions.erase(parentId);
+            }
+        }
+    }
```
Nit: This code block can be extracted in a separate function because after reading the code above that's mostly delegation consecutive function calls, it'd be nice to have something similar here as well - `removeParentTxs`. One more point being that `seen_transactions` variable is internal to this logic and doesn't need to be exposed to `processBlock`.

ismaelsadeeq replied:
Fixed

rkrux commented on src/policy/fees.cpp:
```diff
@@ -690,6 +690,20 @@ void CBlockPolicyEstimator::processBlock(const std::vector<RemovedMempoolTransac
+        seen_transactions.insert(tx.info.m_tx->GetHash());
+        for (const auto& input : tx.info.m_tx->vin) {
+            const Txid& parentId = input.prevout.hash;
+            if (seen_transactions.count(parentId)) {
```
I'm assuming the transactions are sorted from parents to children. So in the case A -> B -> C -> D, all except D would be removed, right?

ismaelsadeeq replied:
>  I'm assuming the transactions are sorted from parents to children.

Yes, they are.

The fee estimator does not take all transaction with parents into account already see
https://github.com/bitcoin/bitcoin/blob/7fcf4e99791ca5be0b068ac03a81a50ece11dba3/src/policy/fees.cpp#L615

Assuming A->B->C->D are topologically sorted from parent to descendant.

In this case B, C, D are already not considered for fee estimation, After this PR A will also be ignored.

rkrux replied:
Hmm I see, as per this comment it says "not dependent on any other transactions in the **mempool**" - does this also apply to other transactions in the **block**?

https://github.com/bitcoin/bitcoin/blob/7fcf4e99791ca5be0b068ac03a81a50ece11dba3/src/policy/fees.cpp#L613

ismaelsadeeq replied:
> does this also apply to other transactions in the block?

I don't fully understand your question. 
The comment you highlighted refers to all incoming transactions into the mempool.

After a new block arrives, the fee estimator listens for notifications of all transactions removed from mempool because of the new block, and updates their tracking starts. All transactions with mempool dependencies that are mined are not tracked initially, they are just going to be skipped.

rkrux commented on test/functional/feature_fee_estimation.py:
```diff
@@ -278,12 +283,16 @@ def sanity_check_rbf_estimates(self, utxos):
             while len(utxos_to_respend) > 0:
                 u = utxos_to_respend.pop(0)
                 tx = make_tx(self.wallet, u, high_feerate)
+                self.confutxo.append({
+                    "txid": tx["txid"],
+                    "vout": 0,
+                    "value": Decimal(tx["tx"].vout[0].nValue) / COIN
+                })
```
Wondering why these calls are not required after every `make_tx` call?

ismaelsadeeq replied:
Because `make_tx` is only constructing the transaction, the transaction created might not be confirmed, so Its inaccurate to add it to confirmed utxo's after constructing the transaction.

Transactions added to the confirmed utxo's in the diff (`sanity_check_rbf_estimates`) are all going to be confirmed eventually so we are accounting for them.

rkrux commented on test/functional/feature_fee_estimation.py:
```diff
@@ -383,6 +401,85 @@ def test_acceptstalefeeestimates_option(self):
+        of their child's feerate.
+        """
+        # The broadcaster and block producer
+        broadcaster = self.nodes[0]
+        miner = self.nodes[1]
```
+1 for expressive node names!

rkrux reviewed (approved):
tACK [c12a677](https://github.com/bitcoin/bitcoin/pull/30079/commits/c12a677cc250608171bc4f6311095b60ba24abab)

Make successful, all functional tests successful.
Mentioned few nits and asked few questions for my clarity.

> The downside of this approach is that transactions that are not CPFP’d will aslo be ignored.

Related to this^, one specific case I can think of that will be affected is of layered transactions that are meant to be confirmed together, example: fanout transactions to increase the utxo count in the wallet and make the utxo value distribution more even, but Idk how frequently these transactions happen.

ismaelsadeeq commented on test/functional/feature_fee_estimation.py:
```diff
@@ -421,15 +518,16 @@ def run_test(self):
-        self.sanity_check_rbf_estimates(self.confutxo + self.memutxo)
+        self.sanity_check_rbf_estimates(self.confutxo)
+
+        self.log.info("Restarting node with fresh estimation")
```
Fixed, thanks

ismaelsadeeq commented on test/functional/feature_fee_estimation.py:
```diff
@@ -383,6 +401,85 @@ def test_acceptstalefeeestimates_option(self):
+        # In sat/vb
+        low_feerate = Decimal(2)
+        med_feerate = Decimal(15)
+        high_feerate = Decimal(20)
```
fixed

ismaelsadeeq commented:
> > Do we have charts anywhere tracking % of transactions that are in a cluster size of 1?
> 
> I will analyze the percentage of cluster size 1 transactions mined in previous blocks.

I tracked recent 1000 blocks from block `842457` to `843457`

~61% of transactions in the last 1000 blocks were confirmed in a cluster size > 1.
~38% of transactions in the last 1000 blocks were confirmed in a cluster size 1

Transactions: 3974143
Cluster size 1 transactions: 1516505
Approximate percentage of Cluster size 1 transactions: ~38
Transactions in a cluster size > 1 transactions: 2457638
Approximate percentage of transactions in a cluster size  > 1: ~61

The data is  available here here https://docs.google.com/spreadsheets/d/1QTE2EQeQlamA123pIn0SY4F-9jtnzS7e7CV1etT6ytM/edit?usp=sharing

I used this script to collect the data https://gist.github.com/ismaelsadeeq/e2767040b720f65986976b3f8c6b7b20

ismaelsadeeq commented:
Thanks for review @rkrux
- I addressed all your comments
- Force pushed from c12a677cc250608171bc4f6311095b60ba24abab to 2563305c0aef3975a6321911db7e0f2a245486de
[compare diff](https://github.com/bitcoin/bitcoin/compare/c12a677cc250608171bc4f6311095b60ba24abab..2563305c0aef3975a6321911db7e0f2a245486de)

rkrux commented on test/functional/feature_fee_estimation.py:
```diff
@@ -383,6 +401,79 @@ def test_acceptstalefeeestimates_option(self):
 
 
+    def send_and_mine_child_tx(self, broadcaster, miner, parent_tx, feerate):
+        u = {"txid": parent_tx["txid"], "vout": 0, "value": Decimal(parent_tx["tx"].vout[0].nValue) / COIN}
```
`Decimal(parent_tx["tx"].vout[0].nValue) / COIN`

Is there not a `satToBtc()` already? WDYT about introducing one? I've noticed this conversion in this diff thrice.

ismaelsadeeq replied:
Good idea for a follow-up @rkrux

josibake replied:
yeah, would recommend this be a separate PR, since there are likely multiple instances in other tests where we could replace the satToBtc conversion with a utility function in the test framework.

rkrux replied:
Yeah, I can pick this up in a separate PR.

rkrux replied:
Quite a late reply here but I don't think I will get to this anytime soon, created an issue here for anyone else to pick up: https://github.com/bitcoin/bitcoin/issues/31345

rkrux reviewed (approved):
reACK [2563305](https://github.com/bitcoin/bitcoin/pull/30079/commits/2563305c0aef3975a6321911db7e0f2a245486de)

Thanks for the quick turnaround on this @ismaelsadeeq!

josibake reviewed (commented):
Looks good, although unclear to me why we don't also remove the child transaction from fee estimation to avoid overestimating. Would recommend also removing the children, or at least documenting why it doesn't matter in a comment.

glozow commented on src/policy/fees.cpp:
```diff
@@ -662,6 +662,22 @@ bool CBlockPolicyEstimator::processBlockTx(unsigned int nBlockHeight, const Remo
     return true;
 }
 
+void CBlockPolicyEstimator::removeParentTxs(const std::vector<RemovedMempoolTransactionInfo>& txs_removed_for_block)
```
> although unclear to me why we don't also remove the child transaction from fee estimation to avoid overestimating

A child is already ignored, new mempool txns with unconfirmed parents are never added

https://github.com/bitcoin/bitcoin/blob/dd42a5ddea6a72e1e9cad54f8352c76b0b701973/src/validation.cpp#L1274-L1279
https://github.com/bitcoin/bitcoin/blob/dd42a5ddea6a72e1e9cad54f8352c76b0b701973/src/policy/fees.cpp#L614-L615

ismaelsadeeq replied:
Thanks for your review, Josie. I think you are missing something.

When the child arrives, it will have a mempool parent, and we currently do not consider all transactions with mempool parent for fee estimation.

So when processing a transaction in `processBlockTx` all transactions with mempool parent, will be missing and , and we will skip them.

Child transactions whose parents aren't in our mempool will also be missing, and we will skip them because orphan transactions aren't added to the mempool and hence won't be tracked. 
They are skipped here:
https://github.com/bitcoin/bitcoin/blob/dd42a5ddea6a72e1e9cad54f8352c76b0b701973/src/policy/fees.cpp#L640

See the discussion with @rkrux:
https://github.com/bitcoin/bitcoin/pull/30079#discussion_r1599598936

I recently made a delving on how fix this limitation and track chunks post cluster mempool.
[Package-aware fee estimator post](https://delvingbitcoin.org/t/package-aware-fee-estimator-post-cluster-mempool/312)

ismaelsadeeq replied:
Added the comment, thanks.

ismaelsadeeq committed:
fees: add `GetMiniMinerInput` to fees

- This method takes removed mempool transaction and generates
  an output that can be linearized by mini miner.

Co-authored-by: willcl-ark <will@256k1.dev>

laib200 commented:
براكة الله فيك 😊

ismaelsadeeq commented on src/policy/fees.h:
```diff
@@ -303,6 +303,9 @@ class CBlockPolicyEstimator : public CValidationInterface
     /** Process a transaction confirmed in a block*/
     bool processBlockTx(unsigned int nBlockHeight, const RemovedMempoolTransactionInfo& tx) EXCLUSIVE_LOCKS_REQUIRED(m_cs_fee_estimator);
 
+    /* Remove transactions with child from fee estimation tracking stats */
```
Fixed

ismaelsadeeq commented:
> > > Do we have charts anywhere tracking % of transactions that are in a cluster size of 1?
> > 
> > 
> > I will analyze the percentage of cluster size 1 transactions mined in previous blocks.
> 
> I tracked recent 1000 blocks from block `842457` to `843457`
> 
> ~61% of transactions in the last 1000 blocks were confirmed in a cluster size > 1. ~38% of transactions in the last 1000 blocks were confirmed in a cluster size 1
> 
> Transactions: 3974143 Cluster size 1 transactions: 1516505 Approximate percentage of Cluster size 1 transactions: ~38 Transactions in a cluster size > 1 transactions: 2457638 Approximate percentage of transactions in a cluster size > 1: ~61
> 
> The data is available here here https://docs.google.com/spreadsheets/d/1QTE2EQeQlamA123pIn0SY4F-9jtnzS7e7CV1etT6ytM/edit?usp=sharing
> 
> I used this script to collect the data https://gist.github.com/ismaelsadeeq/e2767040b720f65986976b3f8c6b7b20


This data shows that most transactions are in cluster size > 1.

So I switched to ignoring transaction that are CPFP'd only.

Thanks for your reviews @rkrux @josibake 
- Forced pushed from 2563305c0aef3975a6321911db7e0f2a245486de to 57556e5e8f11fe3eaee8dfd6a61bcdf60c7e0fbe [Compare diff](https://github.com/bitcoin/bitcoin/compare/2563305c0aef3975a6321911db7e0f2a245486de..57556e5e8f11fe3eaee8dfd6a61bcdf60c7e0fbe)

- Addressed all comments.

josibake commented:
I didn't look too closely into the CI failure reason, but seems like a relatively simple fix, i.e. fixing an signed int overflow: 

```
SUMMARY: UndefinedBehaviorSanitizer: signed-integer-overflow node/mini_miner.cpp:196:33 
MS: 0 ; base unit: 0000000000000000000000000000000000000000
artifact_prefix='./'; Test unit written to ./crash-2315c89458602360eb93362328da5c0ef21f9864

Traceback (most recent call last):
  File "/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/fuzz/test_runner.py", line 411, in <module>
    main()
  File "/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/fuzz/test_runner.py", line 199, in main
    run_once(
  File "/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/fuzz/test_runner.py", line 376, in run_once
    assert len(done_stat) == 1
           ^^^^^^^^^^^^^^^^^^^
AssertionError
```

ismaelsadeeq commented:
Marking as draft, as I attempt to fix the fuzzing overflow error.

ismaelsadeeq commented:
> I didn't look too closely into the CI failure reason, but seems like a relatively simple fix, i.e. fixing an signed int overflow:

@josibake overflow is from mini miner `AncestorFeerateComparator`, added a fixup commit here and opened an issue for it.
https://github.com/bitcoin/bitcoin/issues/30284

rkrux commented on src/test/feesutil_tests.cpp:
```diff
@@ -0,0 +1,328 @@
+        const CTransactionRef txJ = make_tx(outpoints);
+        transactions.emplace_back(entry.FromTx(txJ));
+
+        // Create cluster B child ---> K
```
```
// Create cluster D child ---> K
```

rkrux commented on src/test/feesutil_tests.cpp:
```diff
@@ -0,0 +1,328 @@
+        const CTransactionRef txJ = make_tx(outpoints);
+        transactions.emplace_back(entry.FromTx(txJ));
+
+        // Create cluster B child ---> K
```
```
// Create cluster D child ---> K
```

ismaelsadeeq replied:
fixed

rkrux commented on src/test/feesutil_tests.cpp:
```diff
@@ -0,0 +1,328 @@
+        const auto txAncestorsAndDescendants = GetTxAncestorsAndDescendants(transactions);
+        BOOST_CHECK_EQUAL(txAncestorsAndDescendants.size(), transactions.size());
+        // Check tx A topology;
+        {
+            const Txid txA_Id = transactions[0].info.m_tx->GetHash();
+            const auto txA_Iter = txAncestorsAndDescendants.find(txA_Id);
+            BOOST_CHECK(txA_Iter != txAncestorsAndDescendants.end());
+            const auto ancestors = std::get<0>(txA_Iter->second);
+            BOOST_CHECK(ancestors.size() == 1); // A
+            BOOST_CHECK(ancestors.find(txA_Id) != ancestors.end());
+            const auto descendants = std::get<1>(txA_Iter->second);
+            BOOST_CHECK(descendants.size() == 4); // A, E, F, G
+            BOOST_CHECK(descendants.find(txA_Id) != descendants.end());
+            for (auto i = 4; i <= 6; i++) {
+                auto curr_tx = transactions[i];
+                BOOST_CHECK(descendants.find(curr_tx.info.m_tx->GetHash()) != descendants.end());
+            }
+        }
```
There are 9 such assertion blocks with almost similar validations on ancestors and descendants.
If the author feels it's worth it, IMO this duplication can be eliminated by accepting few variables as the input in a function and calling it with different arguments.
```
validateTxRelations(txToValidate, ancestorTxs, descendantTxs)
```

ismaelsadeeq replied:
thanks accepted your suggestion with a slight modification.

rkrux replied:
Great, thank you!

rkrux commented on src/policy/fees_util.cpp:
```diff
@@ -0,0 +1,31 @@
+#include <policy/fees_util.h>
+#include <primitives/transaction.h>
+
+TxAncestorsAndDescendants GetTxAncestorsAndDescendants(const std::vector<RemovedMempoolTransactionInfo>& transactions)
```
+1 on separating this function out in a file and adding focussed unit tests for it, I feel this is generic enough to be used in several places later.

rkrux commented on src/policy/fees_util.cpp:
```diff
@@ -0,0 +1,31 @@
+    TxAncestorsAndDescendants visited_txs;
+    for (const auto& tx_info : transactions) {
+        const auto& txid = tx_info.info.m_tx->GetHash();
+        std::set<Txid> tx_set{txid};
+        visited_txs.emplace(txid, std::make_tuple(tx_set, tx_set));
```
Interesting that `tx_set` is passed by value here and 2 separate copies are made at the time of calling, which is the need here for separate ancestors and descendants copies.

ismaelsadeeq replied:
yes, this is gone now.

rkrux reviewed (approved):
tACK [cb95f95](https://github.com/bitcoin/bitcoin/pull/30079/commits/cb95f95907cf9fce3a18f2bbbb059e72a897e34e)

`make, test/functional` are successful.

Although, this round of review was focussed on [5e6ff0f](https://github.com/bitcoin/bitcoin/pull/30079/commits/5e6ff0f80e2ed7daadb3abbd5a773b542b9b0f6f)

ismaelsadeeq committed:
mini miner: enable `Linearize` return packages and their selection order

Co-authored-by: willcl-ark <will@256k1.dev>

glozow commented on src/node/mini_miner.cpp:
```diff
@@ -189,11 +189,7 @@ struct AncestorFeerateComparator
-            //     ancestor_fee / ancestor_size <= tx_fee / tx_size
-            // Avoid division and possible loss of precision by
-            // multiplying both sides by the sizes:
-            return ancestor_fee * tx_size < tx_fee * ancestor_size ?
+            return CFeeRate(ancestor_fee, ancestor_size) <= CFeeRate(tx_fee, tx_size) ?
```
instead of b1b28a21e185dfbc0c75f92ac55401a9ba8a5b08, I'd recommend rebasing on #30412
I was able to reproduce the ub crash and verify that 6254c253e995e77196b853784cfd0f42619b324d fixes it.

ismaelsadeeq replied:
thanks I will rebase on it.

ismaelsadeeq replied:
done

ismaelsadeeq commented:
Rebased and added newly introduced files to cmakelist file instead of makefile

Bambibrown commented on src/test/miniminer_tests.cpp:
```diff
@@ -690,7 +690,7 @@ BOOST_FIXTURE_TEST_CASE(manual_ctor, TestChain100Setup)
         node::MiniMiner miniminer_manual(miniminer_info, descendant_caches);
         BOOST_CHECK(miniminer_manual.IsReadyToCalculate());
-        const auto sequences{miniminer_manual.Linearize()};
+        const auto sequences{miniminer_manual.Linearize().inclusion_order};
```
```suggestion
        const auto ### _**sequences{miniminer_manual.Linearize().inclusion_order};**_
```

ismaelsadeeq commented:
@DrahtBot I've rebased and force-pushed 2 hours ago, but the update is yet to reflect here see https://github.com/bitcoin/bitcoin/compare/master...ismaelsadeeq:bitcoin:05-2023-ignore-transactions-with-parents

ismaelsadeeq commented:
> This is a GitHub bug, but they are not going to fix it. The only way to fix it is by doing another rebase (or for a GitHub staff to manually clean it). See [maflcko/DrahtBot#43 (comment)](https://github.com/maflcko/DrahtBot/issues/43#issuecomment-2653314333)

I rebase against master and force-push with newer commits, it is still not updating https://github.com/bitcoin/bitcoin/compare/master...ismaelsadeeq:bitcoin:05-2023-ignore-transactions-with-parents

ismaelsadeeq commented:
I restored the last synced push, but I can not access the re-open button
<img width="1034" alt="Screenshot 2025-03-18 at 15 59 51" src="https://github.com/user-attachments/assets/81c71341-530f-4496-ad88-3552c4d83409" />

yancyribbens commented on src/node/mini_miner.h:
```diff
@@ -162,10 +169,9 @@ class MiniMiner
     /** Construct a new block template with all of the transactions and calculate the order in which
-     * they are selected. Returns the sequence number (lower = selected earlier) with which each
-     * transaction was selected, indexed by txid, or an empty map if it cannot be calculated.
+     * they are selected and the packages fee rate and sizes.
```
The sentence used to end at `they are selected.`  Now it's a very long run on sentence.  It would be better to break it into two sentences or shorten it.

ismaelsadeeq replied:
fixed.

ismaelsadeeq commented:
Rebased and fix comment by @yancyribbens

josibake commented:
Thanks for taking the suggestion of splitting this up! I think it would be nice to add a blurb to the description that this PR depends on https://github.com/bitcoin/bitcoin/pull/33216 and is in draft pending review on the parent PR. You mention this in the comments, but having it in the description ensures reviewers dont miss it / get confused on the status of the PR.
